### PR TITLE
feat: transaction encoding

### PR DESCRIPTION
**Pull request by leoslr**
## Description of the changes

It exports new methods from the `payment-processor` package:
- Encoding of the approval transaction for a request payment
- Check for knowing if an approval is needed for a request payment
- Encoding of an approval transaction if needed for a request payment
- Encoding of the payment transaction for a request payment
- Encoding of all necessary transactions (approval if needed + payment) for a request payment

New tests cover these methods

### CODE REVIEW

File: packages/payment-processor/src/payment/encoder-payment.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  options?: IRequestPaymentOptions,
+): Promise<IPreparedTransaction> {
+  if (options && options.swap) {
+    return encodeRequestPaymentWithSwap(request, provider, options);
+  } else {
+    return encodeRequestPaymentWithoutSwap(request, options);
+  }
+}
+
+export async function encodeRequestPaymentWithoutSwap(
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/encoder-payment.ts:28]: this method (and the next) should not be async

File: packages/payment-processor/src/payment/encoder-approval.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  from: string,
+  options: IRequestPaymentOptions,
+): Promise<IPreparedTransaction | void> {
+  const paymentNetwork = getPaymentNetworkExtension(request)?.id;
+
+  switch (paymentNetwork) {
+    case ExtensionTypes.ID.PAYMENT_NETWORK_ERC20_FEE_PROXY_CONTRACT:
+      if (options && options.swap) {
+        if (
+          !(await hasApprovalErc20ForSwapToPay(
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/encoder-approval.ts:94]: I think it's better to keep the check for approval outside of this. Actually we need a generic top level method to approve if needed. But the encoding should not be concerned by it. It should remain a simple, synchronous, method

File: packages/payment-processor/src/payment/index.ts [hunk 3]
```diff
@@ -116,6 +120,37 @@ export async function payRequest(
...
 
+/**
+ * Encode the transactions associated to a request
+ * @param request the request to pay.
+ * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ * @param from The address which will send the transaction.
+ * @param options encoding options
+ * @returns
+ */
+export async function encodeRequest(
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/index.ts:131]: see my other comment regarding the responsability of the encoding function

name is too generic

```suggestion
export function encodeRequestPayment(
```

**Review by benjlevesque — COMMENTED** [review]
super nice! The encoding should actually be used in the `payRequest` method as well.

File: packages/payment-processor/src/payment/encoder-approval.ts (hunk 2 above)
Comment (leoslr) [reply packages/payment-processor/src/payment/encoder-approval.ts:269]: Now there are the following methods (global, with and without swa):
- Approval encoding
- Check for knowing if approval is needed
- Approval encoding if needed

File: packages/payment-processor/src/payment/index.ts (hunk 3 above)
Comment (leoslr) [reply packages/payment-processor/src/payment/index.ts]: I modified this one to `encodeRequestApprovalAndPayment`. The specific methods for approval or payment encoding can be found in `encoder-approval` and `encoder-payment`.

**Comment by leoslr** [comment]
@benjlevesque I agree that we could use the encoding method inside the `payRequest` method - this way we could also export a global payment method for both 'classic' and swap payment.
However the methods arguments of the encoding method cover a larger scope than the one from the payment method, so changing that would introduce breaking changes for invoicing. 

Maybe it can be covered in a separate PR, what do you think ?

File: packages/payment-processor/src/payment/encoder-payment.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      return {
+        ...prepareEthFeeProxyPaymentTransaction(request, amount, feeAmount),
+        ...overrides,
+      };
+    default:
+      throw new Error('Payment network not found');
+  }
+}
+
+export async function encodeRequestPaymentWithSwap(
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/encoder-payment.ts:101]: remove async?

File: packages/payment-processor/src/payment/encoder-payment.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  options?: IRequestPaymentOptions,
+): Promise<IPreparedTransaction> {
+  if (options && options.swap) {
+    return encodeRequestPaymentWithSwap(request, provider, options);
+  } else {
+    return encodeRequestPaymentWithoutSwap(request, options);
+  }
+}
+
+export async function encodeRequestPaymentWithoutSwap(
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/encoder-payment.ts:28]: remove async?

File: packages/payment-processor/test/payment/encoder.test.ts [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    expect(confirmedTx.status).toBe(1);
+    expect(tx.hash).not.toBeUndefined();
+
+    tx = await wallet.sendTransaction(encodedTransactions[1]);
+    confirmedTx = await tx.wait(1);
+    expect(confirmedTx.status).toBe(1);
+    expect(tx.hash).not.toBeUndefined();
+  });
+
+  it.only('Should handle ERC20 Swap and Conversion Proxy request', async () => {
```
Comment (benjlevesque) [inline packages/payment-processor/test/payment/encoder.test.ts:290]:
```suggestion
  it('Should handle ERC20 Swap and Conversion Proxy request', async () => {
```

**Review by benjlevesque — APPROVED** [review]

**Review by KolevDarko — APPROVED** [review]

**Pull request merged**